cmd/retagger/main.go

```diff
@@ -0,0 +1,245 @@
+	"os"
+	"os/signal"
+	"path"
+	"sync"
+
+	"github.com/containerd/containerd/v2/core/content"
+	"github.com/containerd/containerd/v2/core/remotes"
+	"github.com/containerd/containerd/v2/core/transfer/registry"
+	"github.com/containerd/platforms"
+	"github.com/cpuguy83/dockercfg"
+	"github.com/distribution/reference"
+	"github.com/goccy/go-yaml"
+	ocispec "github.com/opencontainers/image-spec/specs-go/v1"
+)
+
+type retagConfig struct {
+	Source string `json:"source" yaml:"source"`
+	Dest   string `json:"dest" yaml:"dest"`
+}
+
+type runConfig struct {
+	RepoPrefix string
+}
+
+func main() {
+	var rcfg runConfig
+
+	flag.StringVar(&rcfg.RepoPrefix, "retag-registry", os.Getenv("RETAG_REGISTRY"), "Prefix for the repository URLs")
+	flag.Parse()
+
+	config := flag.Arg(0)
+	ctx, cancel := signal.NotifyContext(context.Background(), os.Interrupt)
+	defer cancel()
+
+	if err := run(ctx, config, rcfg); err != nil {
+		fmt.Fprintln(os.Stderr, err)
+		os.Exit(1)
+	}
+}
+
+type errgroupCollector struct {
+	group sync.WaitGroup
+
+	mu   sync.Mutex
+	errs []error
+}
+
+func (ec *errgroupCollector) Do(f func() error) {
+	ec.group.Add(1)
+
+	go func() {
+		defer ec.group.Done()
+
+		if err := f(); err != nil {
+			ec.mu.Lock()
+			defer ec.mu.Unlock()
+			ec.errs = append(ec.errs, err)
+		}
+	}()
+}
+
+func (ec *errgroupCollector) Wait() error {
+	ec.group.Wait()
+
+	ec.mu.Lock()
+	defer ec.mu.Unlock()
+	if len(ec.errs) == 0 {
+		return nil
+	}
+
+	return fmt.Errorf("encountered %d errors: %w", len(ec.errs), errors.Join(ec.errs...))
+}
+
+func run(ctx context.Context, configPath string, rcfg runConfig) error {
+	dt, err := os.ReadFile(configPath)
+	if err != nil {
+		return fmt.Errorf("error reading config file: %w", err)
+	}
+
+	var ls []retagConfig
+	if err := yaml.Unmarshal(dt, &ls); err != nil {
+		return fmt.Errorf("error unmarshalling config: %w", err)
+	}
+
+	var authFn credentialHelperFunc
+	dcfg, err := dockercfg.LoadDefaultConfig()
+	if err != nil {
+		if !errors.Is(err, fs.ErrNotExist) {
+			return fmt.Errorf("error loading Docker config: %w", err)
+		}
+		authFn = func(h string) (string, string, error) {
+			return dockercfg.GetCredentialsFromHelper("", h)
+		}
+	}
+
+	if authFn == nil {
+		authFn = dcfg.GetRegistryCredentials
+	}
+
+	var eg errgroupCollector
+	for _, cfg := range ls {
+		eg.Do(func() (retErr error) {
+			defer func() {
+				if retErr != nil {
+					retErr = fmt.Errorf("error processing %s: %w", cfg.Source, retErr)
+				}
+			}()
+			cfg.Dest = path.Join(rcfg.RepoPrefix, cfg.Dest)
+
+			ref, err := reference.ParseAnyReference(cfg.Source)
+			if err != nil {
+				return fmt.Errorf("error parsing source image %s: %w", cfg.Source, err)
+			}
+
+			src, err := registry.NewOCIRegistry(ctx, ref.String(), registry.WithCredentials(authFn))
+			if err != nil {
+				return fmt.Errorf("error creating source registry for %s: %w", cfg.Source, err)
+			}
+
+			dst, err := registry.NewOCIRegistry(ctx, cfg.Dest, registry.WithDefaultScheme("http"))
```

Comment:
Should WithCredentials happen here too? Or is authentication happening elsewhere?

Reply:
Yep, good catch.
Also the default scheme is not what I want either.